lib/io/can.cpp

```diff
@@ -1,10 +1,124 @@
+    return core::Result::kFailure;
+  }
+  const int num_bytes_written = write(socket_, &data, sizeof(CanFrame));
+  if (num_bytes_written != sizeof(CanFrame)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to send CAN data");
+    return core::Result::kFailure;
+  }
+  // TODOLater make logger_ more elegant
+  logger_.log(core::LogLevel::kDebug,
+              "CAN data sent, ID:%i, DATA: %i %i %i %i %i %i %i %i",
+              static_cast<int>(data.can_id),
+              static_cast<int>(data.data[0]),
+              static_cast<int>(data.data[1]),
+              static_cast<int>(data.data[2]),
+              static_cast<int>(data.data[3]),
+              static_cast<int>(data.data[4]),
+              static_cast<int>(data.data[5]),
+              static_cast<int>(data.data[6]),
+              static_cast<int>(data.data[7]));
+  return core::Result::kSuccess;
+}
+
+std::optional<CanFrame> Can::receive()
+{
+  CanFrame received_message;
+  if (ioctl(socket_, FIONREAD) < sizeof(CanFrame)) {
+    logger_.log(core::LogLevel::kDebug, "No CAN data in rx queue");
+    return std::nullopt;
+  }
+  const int num_bytes_read = read(socket_, &received_message, sizeof(CanFrame));
+  if (num_bytes_read < sizeof(CanFrame)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to receive CAN data");
+    return std::nullopt;
+  }
+  // TODOLater make logger more elegant
+  logger_.log(core::LogLevel::kDebug,
+              "CAN message received, ID:%i, DATA: %i %i %i %i %i %i %i %i",
+              static_cast<int>(received_message.can_id),
+              static_cast<int>(received_message.data[0]),
+              static_cast<int>(received_message.data[1]),
+              static_cast<int>(received_message.data[2]),
+              static_cast<int>(received_message.data[3]),
+              static_cast<int>(received_message.data[4]),
+              static_cast<int>(received_message.data[5]),
+              static_cast<int>(received_message.data[6]),
+              static_cast<int>(received_message.data[7]));
+
+  return received_message;
+}
+
+core::Result Can::listen()
+{
+  if (ioctl(socket_, FIONREAD) < sizeof(CanFrame)) { return core::Result::kFailure; }
+  const auto message = receive();
+  if (!message) { return core::Result::kFailure; }
+  const auto subscribed_processors = processors_.find(message->can_id);
+  if (subscribed_processors == processors_.end()) {
+    logger_.log(core::LogLevel::kInfo, "No CanProccessor associated with id %i", message->can_id);
+    return core::Result::kFailure;
+  }
+  for (auto &processor : subscribed_processors->second) {
+    processor->processMessage(*message);
+  }
+  return core::Result::kSuccess;
+}
+
+void Can::addCanProcessor(const std::uint16_t id, std::shared_ptr<ICanProcessor> processor)
 {
-  // TODOLater implement
+  const auto id_and_processors = processors_.find(id);
+  if (id_and_processors == processors_.end()) {
+    processors_.emplace(id, std::vector<std::shared_ptr<ICanProcessor>>({processor}));
```

Comment:
Not a fan of this line. Can you split this up nicely in order to avoid complexity?